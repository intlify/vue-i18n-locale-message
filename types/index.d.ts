/**
 *  Locale Message Recursive Structure
 *    e.g.
 *    {
 *      "en": {  // 'en' Locale
 *        "key1": "this is message1", // basic
 *        "nested": { // nested
 *          "message1": "this is nested message1"
 *        },
 *        "errors": [ // array
 *          "this is 0 error code message",
 *          {  // object in array
 *            "internal1": "this is internal 1 error message"
 *          },
 *          [  // array in array
 *            "this is nested array error 1"
 *          ]
 *        ]
 *      },
 *      "ja": { // 'ja' Locale
 *        // ...
 *      }
 *    }
 */

export type Locale = string
export type LocaleMessage =
  | string
  | { [property: string]: LocaleMessage }
  | LocaleMessage[]
export type LocaleMessages = Record<Locale, LocaleMessage>

/**
 *  Locale Message Meta Structure
 *    e.g.
 *    {
 *     "target": "/path/to/project1",
 *     "components": { // flat component paths
 *       "/path/to/project1/src/App.vue": [ // use array for mutliple i18n blocks
 *         {
 *           "lang": "json", // or "yaml", "json5"
 *           "messages": { // messages
 *             "ja": { // languages
 *               "title": "アプリケーション",
 *               "lang": "言語切り替え"
 *             },
 *             "en": {
 *               "title": "Application",
 *               "lang": "Change languages"
 *             }
 *           }
 *         }
 *       ],
 *       "/path/to/project1/src/components/Modal.vue": [
 *         {
 *           "lang": "yaml",
 *           "locale": "ja",
 *           "messages": {
 *             "ja": {
 *               "ok": "OK",
 *               "cancel": "キャンセル"
 *             }
 *           }
 *         }
 *       ]
 *     }
 *   }
 */

export type I18nLang = 'json' | 'yaml' | 'yml' | 'json5'
export type SFCI18nBlock = {
  lang: I18nLang,
  locale?: string,
  src?: string,
  messages: LocaleMessages
}
export type MetaLocaleMessage = {
  target: string,
  components: Record<string, SFCI18nBlock[]>
}
export type FormatOptions = {
  intend?: number,
  eof?: string
}

/**
 *  SFC (Single-file component) file info
 *    e.g.
 *    {
 *      path: '/path/to/project1/src/components/common/Modal.vue',
 *      content: `
 *        <template>
 *          <!-- template contents is here ... -->
 *        </template>
 *
 *        <script>
 *          // script codes is here ...
 *        </script>
 *
 *        <style scoped>
 *          // css style codes is here ...
 *        </style>
 *
 *        <i18n>
 *        {
 *          "en": {
 *            "ok": "OK",
 *            "cancel": "Cancel"
 *          },
 *          "ja": {
 *            "ok": "OK",
 *            "cancel": "キャンセル"
 *          }
 *        }
 *        </i18n>
 *      `,
 *    }
 */

export interface SFCFileInfo {
  path: string
  content: string
}

/**
 *  Extend SFCDescriptor, due to squeeze / infuse.
 *    e.g.
 *    {
 *      template: { ... },
 *      script: { ... },
 *      styles: [{ ... }],
 *      customBlocks: [{
 *        type: 'i18n',
 *        content: `
 *        {
 *          "en": {
 *            "ok": "OK",
 *            "cancel": "Cancel"
 *          },
 *          "ja": {
 *            "ok": "OK",
 *            "cancel": "キャンセル"
 *          }
 *        }
 *        `,
 *        attrs: { ... },
 *        start: 10,
 *        end: 30,
 *        map: { ... }
 *      }, ...],
 *      contentPath: '/path/to/project1/src/components/common/Modal.vue',
 *      component: 'Modal',
 *      hierarchy: ['components', 'common', 'Modal']
 *    }
 */

declare function squeeze (basePath: string, files: SFCFileInfo[]): MetaLocaleMessage
declare function infuse (basePath: string, sources: SFCFileInfo[], meta: MetaLocaleMessage, options?: FormatOptions): SFCFileInfo[]

export type ProviderPushMode = 'file-path' | 'locale-message'

export type ProviderPushFileInfo = {
  locale: Locale
  path: string
}

export type ProviderPushResource = {
  mode: ProviderPushMode
  files?: ProviderPushFileInfo[]
  messages?: LocaleMessages
}

export type ProviderPullResource = LocaleMessages

export interface Provider {
  push (resource: ProviderPushResource, dryRun: boolean): Promise<void>
  pull (locales: Locale[], dryRun: boolean): Promise<ProviderPullResource>
}

export type ProviderFactory<T = {}> = (configration: ProviderConfiguration<T>) => Provider

/**
 *  ProviderConfiguration provider fields structure
 *    e.g.
 *    {
 *      "provider": {
 *        "token": "xxx"
 *      },
 *      "pushMode": "file-path"
 *    }
 */

export interface ProviderConfiguration<T = {}> {
  provider: { [key in keyof ProviderConfigurationValue<T>]: ProviderConfigurationValue<T>[key] }
  pushMode: ProviderPushMode
}

export type ProviderConfigurationValue<T = {}> = T & { [prop: string]: unknown }
