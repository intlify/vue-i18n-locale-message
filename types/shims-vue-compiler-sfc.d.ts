import { SFCDescriptor } from '@vue/compiler-sfc'

// extend for vue-i18n-locale-message
declare module '@vue/compiler-sfc' {
  export interface SFCDescriptor {
    raw: string
    contentPath: string
    component: string
    hierarchy: string[]
  }
}
