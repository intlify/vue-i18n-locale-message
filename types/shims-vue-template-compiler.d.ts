import { SFCDescriptor } from 'vue-template-compiler'

// extend for vue-i18n-locale-message
declare module 'vue-template-compiler/types/index' {
  export interface SFCDescriptor {
    raw: string
    contentPath: string
    component: string
    hierarchy: string[]
  }
}
