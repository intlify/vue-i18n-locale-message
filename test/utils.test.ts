import { SFCFileInfo, NamespaceDictionary } from '../types'
import deepmerge from 'deepmerge'

import jsonFiles from './fixtures/file/json'
import jsonMetaInfo from './fixtures/meta/json'
import externalLocaleMessages from './fixtures/external'
import squeezeLocaleMessages from './fixtures/squeeze.json'

import {
  reflectSFCDescriptor,
  getTranslationStatus,
  getExternalLocaleMessages,
  splitLocaleMessages
} from '../src/utils'

// ------
// mocks

// l10n service provider module
const mockStatus = jest.fn()
jest.mock('@scope/l10n-service-provider', () => {
  return jest.fn().mockImplementation(() => {
    return { status: mockStatus }
  })
})
import L10nServiceProvider from '@scope/l10n-service-provider'

test('reflectSFCDescriptor', () => {
  const descriptors = reflectSFCDescriptor('/path/to/project1/src', jsonFiles as SFCFileInfo[])
  expect(descriptors[0].contentPath).toEqual(jsonMetaInfo[0].contentPath)
  expect(descriptors[0].component).toEqual(jsonMetaInfo[0].component)
  expect(descriptors[0].hierarchy).toEqual(jsonMetaInfo[0].hierarchy)
})

test('getTranslationStatus: success', async () => {
  // setup mocks
  const mockStatusValue = [{
    locale: 'en',
    percentage: 24.2
  }]
  mockStatus.mockImplementation(({ locales }) => Promise.resolve(mockStatusValue))

  // run
  const status = await getTranslationStatus({
    provider: '@scope/l10n-service-provider',
    conf: './test/fixtures/conf/l10n-service-provider-conf.json'
  })

  expect(status).toMatchObject(mockStatusValue)
})

test('getTranslationStatus: provider not found', async () => {
  // run
  await expect(getTranslationStatus({
    provider: './404-provider.js'
  })).rejects.toThrow('Not found ./404-provider.js provider')
})

test('getExternalLocaleMessages: basic usage', async () => {
  const namespaces: NamespaceDictionary = {
    './test/fixtures/packages/package1/locales/**/*.json': 'package1',
    './test/fixtures/packages/package2/locales/**/*.json': 'package2'
  }
  const withBundle = './test/fixtures/packages/package1/locales/**/*.json,./test/fixtures/packages/package2/locales/**/*.json'
  const withBundleMatch = '([\\w]*)/([\\w]*)\\.json$'
  const messages = getExternalLocaleMessages(namespaces, withBundle, withBundleMatch)

  expect(messages).toMatchSnapshot()
})

test('getExternalLocaleMessages: no namespace', async () => {
  const withBundle = './test/fixtures/packages/package1/locales/**/*.json,./test/fixtures/packages/package2/locales/**/*.json'
  const withBundleMatch = '([\\w]*)/([\\w]*)\\.json$'
  const messages = getExternalLocaleMessages({}, withBundle, withBundleMatch)

  expect(messages).toMatchSnapshot()
})

test('getExternalLocaleMessages: no filename', async () => {
  const withBundle = './test/fixtures/packages/package1/locales/**/*.json,./test/fixtures/packages/package2/locales/**/*.json'
  const withBundleMatch = '([\\w]*)/[\\w]*\\.json$'
  const messages = getExternalLocaleMessages({}, withBundle, withBundleMatch)

  expect(messages).toMatchSnapshot()
})

test('getExternalLocaleMessages: no bundle option', async () => {
  const namespaces: NamespaceDictionary = {
    './test/fixtures/packages/package1/locales/**/*.json': 'package1',
    './test/fixtures/packages/package2/locales/**/*.json': 'package2'
  }
  const messages = getExternalLocaleMessages(namespaces)

  expect(messages).toMatchObject({})
})

test('splitLocaleMessages: basic usage', async () => {
  const messages = deepmerge(squeezeLocaleMessages, externalLocaleMessages)
  const namespaces: NamespaceDictionary = {
    './test/fixtures/packages/package1/locales/**/*.json': 'package1',
    './test/fixtures/packages/package2/locales/**/*.json': 'package2'
  }
  const unbundleTo = './test/fixtures/packages/package1/locales/**/*.json,./test/fixtures/packages/package2/locales/**/*.json'
  const unbundleMatch = '([\\w]*)/([\\w]*)\\.json$'
  const { sfc, external } = splitLocaleMessages(messages, namespaces, unbundleTo, unbundleMatch)

  expect(sfc).toEqual(squeezeLocaleMessages)
  external.forEach(ex => {
    expect(ex.messages).toMatchSnapshot()
  })
})

test('splitLocaleMessages: no bundle option', async () => {
  const messages = deepmerge(squeezeLocaleMessages, externalLocaleMessages)
  const namespaces: NamespaceDictionary = {
    './test/fixtures/packages/package1/locales/**/*.json': 'package1',
    './test/fixtures/packages/package2/locales/**/*.json': 'package2'
  }
  const { sfc, external } = splitLocaleMessages(messages, namespaces)

  expect(sfc).toEqual(messages)
  expect(external).toBeUndefined()
})

test('splitLocaleMessages: no namespace', async () => {
  const messages = deepmerge(squeezeLocaleMessages, externalLocaleMessages)
  const namespaces: NamespaceDictionary = {}
  const { sfc, external } = splitLocaleMessages(messages, namespaces)

  expect(sfc).toEqual(messages)
  expect(external).toBeUndefined()
})
