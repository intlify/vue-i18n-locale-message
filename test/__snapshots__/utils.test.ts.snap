// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExternalLocaleMessages: basic usage 1`] = `
Object {
  "en": Object {
    "package1": Object {
      "common": Object {
        "navigation": Object {
          "title": "title",
        },
      },
    },
    "package2": Object {
      "profile": Object {
        "display_name": "Display Name",
        "email": "E-Mail",
      },
    },
  },
  "ja": Object {
    "package1": Object {
      "common": Object {
        "navigation": Object {
          "title": "タイトル",
        },
      },
    },
    "package2": Object {
      "profile": Object {},
    },
  },
}
`;

exports[`getExternalLocaleMessages: no filename 1`] = `
Object {
  "en": Object {
    "display_name": "Display Name",
    "email": "E-Mail",
    "navigation": Object {
      "title": "title",
    },
  },
  "ja": Object {
    "navigation": Object {
      "title": "タイトル",
    },
  },
}
`;

exports[`getExternalLocaleMessages: no namespace 1`] = `
Object {
  "en": Object {
    "common": Object {
      "navigation": Object {
        "title": "title",
      },
    },
    "profile": Object {
      "display_name": "Display Name",
      "email": "E-Mail",
    },
  },
  "ja": Object {
    "common": Object {
      "navigation": Object {
        "title": "タイトル",
      },
    },
    "profile": Object {},
  },
}
`;

exports[`splitLocaleMessages: basic usage 1`] = `
Object {
  "navigation": Object {
    "title": "title",
  },
}
`;

exports[`splitLocaleMessages: basic usage 2`] = `
Object {
  "navigation": Object {
    "title": "タイトル",
  },
}
`;

exports[`splitLocaleMessages: basic usage 3`] = `
Object {
  "display_name": "Display Name",
  "email": "E-Mail",
}
`;

exports[`splitLocaleMessages: basic usage 4`] = `Object {}`;
